### PR TITLE
Refactor G2 module

### PR DESCRIPTION
**Pull request by b13decker**
To split out and abstract the polynomial field extension representation. With these commits, we now have an FQP module which will support the needed Fq12 elements.

### CODE REVIEW

File: spec/Spec/BlsEC/FQP.cry [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+ * ```repl
+ * :prove G2_x ~*~ G2_x == G2_xSquared
+ * :prove G2_x ~*~ G2_x ~*~ G2_x == G2_xCubed
+ * ```
+ */
+(~*~) : {n} (fin n, n >= 2) => FQP n -> FQP n -> FQP n
+(~*~) coeffs1 coeffs2 = reduce_while b (2 * `(n) - 1) where
+    b = poly_multi coeffs1 coeffs2
+
+    reduce_while : {a, n} (fin n, n >= 2, fin a, a == 2 * n - 1) => [a](Z Fq) -> Integer -> [n](Z Fq)
```
Comment (mccleeary-galois) [inline spec/Spec/BlsEC/FQP.cry:62]: I believe Cryptol prefers you to not have the same variable for type parameters in helper functions

File: spec/Spec/BlsEC/FQP.cry [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+ * A polynomial field extension type contains a list of
+ *  coefficients which are elements of Fq.
+ */
+type FQP n = [n](Z Fq)
+
+/**
+ * A G2 point is comprised of two Fq2 elements, or sets of coefficients, each of degree two.
+ * Each element is comprised of a real and imaginary coeffsicient.
+ */
+type Fq2 = FQP 2
```
Comment (mccleeary-galois) [inline spec/Spec/BlsEC/FQP.cry:21]: https://github.com/GaloisInc/ckzg-eip-4844-verification/blob/633cc70583bb6c6c195e8770e9d9fa3d3eabc13b/spec/Spec/BlsEC/G2.cry#L52

Line 52 was broken by this change from an enum to to Fq2 it still has the real and imaginary bit

Comment (mccleeary-galois) [reply spec/Spec/BlsEC/FQP.cry:21]: Also broke this test in Poly.cry
https://github.com/GaloisInc/ckzg-eip-4844-verification/blob/633cc70583bb6c6c195e8770e9d9fa3d3eabc13b/spec/Spec/BlsEC/Poly.cry#L15

File: spec/Spec/BlsEC/Poly.cry [hunk 3]
```diff
@@ -109,29 +108,25 @@ inv_nested_for lm low nm new r = (nm', new') where
...
 inv_nested_for_compute_value : {n} (fin n, n >= 1) => Z Fq -> Integer -> [n](Z Fq) -> [n](Z Fq) -> Z Fq
 inv_nested_for_compute_value ni i l r =
-    foldl (-) ni (map (inv_nested_for_multi l r) (inv_nested_for_index_pairs`{n} i))
+    foldl (-) ni (map (poly_pair_multi l r) (inv_nested_for_index_pairs`{n} i))
 
 /**
- * This function is used by the `foldl` in `inv_nested_for_compute_value` to handle the pairs (-1, -1).
+ * This function performs pair-wise multiplication between two polynomials.
  * ```repl
  * :prove inv_nested_for_multi [13] [13] (-1, 0) == 0
```
Comment (mccleeary-galois) [inline spec/Spec/BlsEC/Poly.cry:116]: Make sure to change these to match new function name

**Review by mccleeary-galois — APPROVED** [review]
LGTM other than comments noted

File: spec/Spec/BlsEC/FQP.cry (hunk 1 above)
Comment (b13decker) [reply spec/Spec/BlsEC/FQP.cry:62]: Even though the two should be the same? That's what I want to indicate with reusing `n`.

Comment (b13decker) [reply spec/Spec/BlsEC/FQP.cry:62]: Oh, yeah that doesn't even work --- apparently I forgot to rerun the tests. Yikes.

**Pull request merged**